### ISSUE TITLE
[Enhancement] Remove STL as possible

### ISSUE THREAD

**Issue by matyalatte** [libui-ng/libui-ng#247]
[CONTRIBUTING.md](https://github.com/matyalatte/libui-ng/blob/master/CONTRIBUTING.md#windows-specific-notes) says we should not use ATL, MFC and WTL.
But I think we should add STL to them because libui is a C library.

I know it's hard to remove some features like `std::vector`,  and maybe we should not remove them.
But we should refrain from using them as possible.

I made a branch as an example.
https://github.com/matyalatte/libui-ng/tree/remove_sstream
I just removed `std::ostreamstring` and `std::string` but the artifacts became much smaller.

|  | Mingw-x64-static-release | Win-x64-static-debug | Win-x86-static-debug |
| :--: | :--: | :--: | :--: |
| [master](https://github.com/matyalatte/libui-ng/actions/runs/6280998076) | 47.2 MB | 140 MB | 135 MB |
| [remove_sstream](https://github.com/matyalatte/libui-ng/actions/runs/6748896696) | 27.9 MB | 117 MB | 112 MB |

**Comment by szanni** [comment]
I like the idea in principle but sadly don't see this happening any time soon.

I have been contemplating moving many of the windows C++ files back to C. You can see in the git log that this used to be the case for many files. My motivation would be to cut down on compile time (probably an order of magnitude reduction). That would require removing all of the C++ stuff from the headers, which turns out to be mainly STL.

As you noted: removing `std::vector` is probably indeed going to be a pain point. Especially in uiGroup and similar.
There are C libraries out there for C vectors, my favorite ones unsurprisingly being my own [vecptr](https://github.com/szanni/vecptr) and [vector](https://github.com/szanni/vector).

Most are macro based and hence SCREEM_AT_YOU_CONSTANTLY or loose type safety, which IMO is a lot worse (e.g. `g_list`). It looks like that will be fixable with C23's `typeof` but that is nothing we will be able to use any time soon, if ever.

I am somewhat surprised by your binary sizes? Are you referring to the entire build directory? Because for me the `tester` binary (which includes most of the ui components) comes out at:
- 13M: static, debug mingw-w64 (Linux)
- 13M: static, release mingw-w64 (Linux)
- 511K: static, release MSVC 2019 (Windows 7 built with the #113 patch)
- 1.7M: static, debug MSVC 2019 (Windows 7 built with the #113 patch)

Especially the MSVC builds seem very reasonable in size to me, even with STL.

Testing your branch gives me 405K (static, release, MSVC 2019) so a 106K reduction. Removing the entire STL would probably get that down even more.
I do love smaller binaries and hate the modern age bloat of software... but I would rate shaving off a few hundred KBs off the binary size to be low priority. Especially if this means increased maintenance overhead due to unwieldy `std::vector` replacements (I would need to see what that actually looks like).

**Comment by bcampbell** [comment]
I'm a little surprised to see that the original switch to C++ and stl was done by @andlabs. I generally respect his tastes and the decisions he made, but stl use does feel a little against the grain... My guess is that he weighed it up and decided the tradeoff was worth it. Maybe he figured that if you were using windows, you'd already lost ;-)
I think that it would be nice to ditch stl use, but I kind of lean toward thinking @andlabs probably made the right decision.

**Comment by bcampbell** [comment]
That said, I think most of the stl use could be ditched pretty easily.
I just had a quick go at getting rid of the std::vector in `windows/uiGrid.cpp`, and the changes were pretty minor. I'd even say it's simpler to do it all in-line rather than using a library or macros - there's usually only one "add" function which has to deal with growing the array, so it's pretty simple to write it bespoke for each control type.
Also a `uiprivRealloc()` would be an easy way simplify things.

In uiGrid, there's also an `std::map` which would be trickier to convert. It's used for the `uiGridInsertAt()`. But to be honest, for `uiGrid`, I think you'd have to have a damn big grid before a std::map lookup was significantly quicker than a brute force linear search...

Anyway here's my uiGrid `std::vector` surgery (caveat - I haven't even bothered booting into windows to compile this!)
```
diff --git a/windows/grid.cpp b/windows/grid.cpp
index c902c71b..068df74b 100644
--- a/windows/grid.cpp
+++ b/windows/grid.cpp
@@ -28,7 +28,9 @@ struct gridChild {
 struct uiGrid {
 	uiWindowsControl c;
 	HWND hwnd;
-	std::vector<struct gridChild *> *children;
+	struct gridChild** children;
+	int children_size;
+	int children_capacity;
 	std::map<uiControl *, size_t> *indexof;
 	int padded;
 
@@ -40,7 +42,8 @@ static bool gridRecomputeMinMax(uiGrid *g)
 {
 	bool first = true;
 
-	for (struct gridChild *gc : *(g->children)) {
+	for (int = 0; i < g->children_size; ++i ) {
+		gridChild* gc = g->children[i];
 		// this is important; we want g->xmin/g->ymin to satisfy gridLayoutData::visibleRow()/visibleColumn()
 		if (!uiControlVisible(gc->c))
 			continue;
@@ -96,7 +99,7 @@ public:
 				this->gg[y][x] = -1;
 		}
 
-		for (i = 0; i < g->children->size(); i++) {
+		for (i = 0; i < g->children_size; i++) {
 			struct gridChild *gc;
 
 			gc = (*(g->children))[i];
@@ -198,7 +201,7 @@ static void gridRelayout(uiGrid *g)
 	struct gridChild *gc;
 	int nhexpand, nvexpand;
 
-	if (g->children->size() == 0)
+	if (g->children_size == 0)
 		return;		// nothing to do
 
 	uiWindowsEnsureGetClientRect(g->hwnd, &r);
@@ -238,7 +241,7 @@ static void gridRelayout(uiGrid *g)
 
 	// 2) figure out which rows/columns expand but not span
 	// we need to know which expanding rows/columns don't span before we can handle the ones that do
-	for (i = 0; i < g->children->size(); i++) {
+	for (i = 0; i < g->children_size; i++) {
 		gc = (*(g->children))[i];
 		if (!uiControlVisible(gc->c))
 			continue;
@@ -250,7 +253,7 @@ static void gridRelayout(uiGrid *g)
 
 	// 3) figure out which rows/columns expand that do span
 	// the way we handle this is simple: if none of the spanned rows/columns expand, make all rows/columns expand
-	for (i = 0; i < g->children->size(); i++) {
+	for (i = 0; i < g->children_size; i++) {
 		gc = (*(g->children))[i];
 		if (!uiControlVisible(gc->c))
 			continue;
@@ -301,7 +304,7 @@ static void gridRelayout(uiGrid *g)
 			ld->rowheights[i] = height / nvexpand;
 
 	// 5) reset the final coordinates for the next step
-	for (i = 0; i < g->children->size(); i++) {
+	for (i = 0; i < g->children_size; i++) {
 		gc = (*(g->children))[i];
 		if (!uiControlVisible(gc->c))
 			continue;
@@ -363,7 +366,7 @@ static void gridRelayout(uiGrid *g)
 
 	// 7) everything as it stands now is set for xalign == Fill yalign == Fill; set the correct alignments
 	// this is why we saved minwidth/minheight above
-	for (i = 0; i < g->children->size(); i++) {
+	for (i = 0; i < g->children_size; i++) {
 		gc = (*(g->children))[i];
 		if (!uiControlVisible(gc->c))
 			continue;
@@ -413,13 +416,14 @@ static void uiGridDestroy(uiControl *c)
 {
 	uiGrid *g = uiGrid(c);
 
-	for (struct gridChild *gc : *(g->children)) {
+	for (int i = 0; i < g->children_size; ++i) {
+		struct gridChild *gc = g->children[i];
 		uiControlSetParent(gc->c, NULL);
 		uiControlDestroy(gc->c);
 		uiprivFree(gc);
 	}
 	delete g->indexof;
-	delete g->children;
+	uiprivFree(g->children);
 	uiWindowsEnsureDestroyWindow(g->hwnd);
 	uiFreeControl(uiControl(g));
 }
@@ -441,8 +445,10 @@ static void uiGridSyncEnableState(uiWindowsControl *c, int enabled)
 
 	if (uiWindowsShouldStopSyncEnableState(uiWindowsControl(g), enabled))
 		return;
-	for (const struct gridChild *gc : *(g->children))
+	for (int i = 0; i < g->children_size; ++i) {
+		const struct gridChild *gc = g->children[i];
 		uiWindowsControlSyncEnableState(uiWindowsControl(gc->c), enabled);
+	}
 }
 
 uiWindowsControlDefaultSetParentHWND(uiGrid)
@@ -460,7 +466,7 @@ static void uiGridMinimumSize(uiWindowsControl *c, int *width, int *height)
 
 	*width = 0;
 	*height = 0;
-	if (g->children->size() == 0)
+	if (g->children_size == 0)
 		return;		// nothing to do
 
 	gridPadding(g, &xpadding, &ypadding);
@@ -533,13 +539,13 @@ static void gridArrangeChildren(uiGrid *g)
 	int i;
 	struct gridChild *gc;
 
-	if (g->children->size() == 0)
+	if (g->children_size == 0)
 		return;		// nothing to do
 	ld = new gridLayoutData(g);
 	controlID = 100;
 	insertAfter = NULL;
-	visited = new bool[g->children->size()];
-	ZeroMemory(visited, g->children->size() * sizeof (bool));
+	visited = new bool[g->children_size];
+	ZeroMemory(visited, g->children_size * sizeof (bool));
 	for (y = 0; y < ycount(g); y++)
 		for (x = 0; x < xcount(g); x++) {
 			i = ld->gg[y][x];
@@ -578,8 +584,22 @@ static void add(uiGrid *g, struct gridChild *gc)
 {
 	uiControlSetParent(gc->c, uiControl(g));
 	uiWindowsControlSetParentHWND(uiWindowsControl(gc->c), g->hwnd);
-	g->children->push_back(gc);
-	(*(g->indexof))[gc->c] = g->children->size() - 1;
+
+	if (g->children_size == g->children_capacity) {
+		// grow (a uiprivRealloc would be nice here...)
+		int newcap = g->children_capacity*2;
+		struct gridChild **newchildren = uiprivAlloc(sizeof(gridChild*) * g->children_capacity, "gridChild** (children)";
+		for (int i = 0; i < g->children_size; ++i) {
+			newchildren[i] = g->children[i];
+		}
+		uiprivFree(g->children);
+		g->children = newchildren;
+		g->children_capacity = newcap;
+	}
+	g->children[g->children_size] = gc;
+	++g->children_size;
+
+	(*(g->indexof))[gc->c] = g->children_size - 1;
 	gridRecomputeMinMax(g);
 	gridArrangeChildren(g);
 	uiWindowsControlMinimumSizeChanged(uiWindowsControl(g));
@@ -649,7 +669,10 @@ uiGrid *uiNewGrid(void)
 
 	g->hwnd = uiWindowsMakeContainer(uiWindowsControl(g), onResize);
 
-	g->children = new std::vector<struct gridChild *>;
+	g->children_size = 0;
+	g->children_capacity = 16;
+	g->children = uiprivAlloc(sizeof(gridChild*) * g->children_capacity, "gridChild** (children)";
+
 	g->indexof = new std::map<uiControl *, size_t>;
 
 	return g;


```

**Comment by cody271** [comment]
> I like the idea in principle but sadly don't see this happening any time soon.
>
> Edit: Just to be clear, I'm not trying to discourage you from embarking on such a mission. Looking at includes like `map` and `vector` however makes me question if spending time on this is actually time well spent.
>
Agreed. Windows code could be rewritten to use a lot less C++, but this is a *huge* amount of work! For not very much gains..

Far more important things missing in the library still.

**Comment by matyalatte** [comment]
Thanks for replies.
I understood why we need `vector` and `map`.
And that's why I wrote "as possible."

I don’t think we should remove STL completely.
But it (or some STL modules) should be deprecated basically.

@szanni said [my commit](https://github.com/matyalatte/libui-ng/commit/35d663a1386b275ef3268feab10730ad49b8a51f) reduced the `tester` binary from 511K to 405K.
But I just rewrote three small functions.
It means **only three functions takes 20% of the binary.**
I don’t think the 20% is negligibly small.
And this might happen when someone uses other STL modules in the future.

@szanni
> I am somewhat surprised by your binary sizes? Are you referring to the entire build directory?

The artifacts means the files uploaded by CI workflows.
You can download them from the links in the table in my first comment.

@cody271
> Far more important things missing in the library still.

True. I don't mind if no one works on it soon.
and I'll respect your decision.

**Comment by rubyFeedback** [comment]
Is it possible to do any change here in partial steps? E. g. remove use of STL in some .c files (or cpp files) but not all?

Anyway. I think this is not really the biggest concern for libui-ng. I have been moving towards Java for the time being, in regards to GUIs - not that I plan to abandon libui-ng, but I need slightly better widget control, and as much as Java SWING sucks, for prototyping it is super great (I use it in combination with jruby actually, but my long term goal is to also tie into GraalVM, once they support Swing more, as a basic stepping stone here; the ultimate goal is to have some kind of DSL, similar to what Andy is doing with the glimmer suite, but probably in a much more agnostic manner than a DSL, e. g. where common components can be described in some text file or so by everyone, rather than the widget layout be described in minute details, which is quite cumbersome.)

**Comment by kojix2** [comment]
The key point here is that removing the STL can significantly reduce the size of the library. I believe this is an important discussion in the context of portable libraries like libui-ng. Thank you RubyFeedback, but the topic of Glimmer is interesting to us Rubyists, but the conversation is getting sidetracked, so let's talk about it elsewhere...
